src/machines/workflow_machines.rs

```diff
@@ -668,14 +674,63 @@ impl WorkflowMachines {
                         .insert(CommandID::ChildWorkflowStart(seq), child_workflow.machine);
                     self.current_wf_task_commands.push_back(child_workflow);
                 }
-                WFCommand::RequestCancelChildWorkflow(RequestCancelExternalWorkflowExecution {
-                    seq: Some(seq),
-                    ..
-                }) => self.process_cancellation(&CommandID::ChildWorkflowStart(seq), &mut jobs)?,
-                WFCommand::RequestCancelChildWorkflow(RequestCancelExternalWorkflowExecution {
-                    seq: None,
-                    ..
-                }) => {}
+                WFCommand::RequestCancelExternalWorkflow(attrs) => match attrs.target {
+                    None => {
+                        return Err(WFMachinesError::Fatal(
+                            "Cancel external workflow command had empty target field".to_string(),
+                        ))
+                    }
+                    Some(cancel_we::Target::ChildWorkflowSeq(seq)) => {
+                        jobs.extend(self.process_cancellation(CommandID::ChildWorkflowStart(seq))?)
+                    }
+                    Some(cancel_we::Target::WorkflowExecution(_)) => {
+                        todo!("Cancel external non-child workflows")
+                    }
+                },
+                WFCommand::SignalExternalWorkflow(attrs) => {
+                    let (we, only_child) = match attrs.target {
+                        None => {
+                            return Err(WFMachinesError::Fatal(
+                                "Signal external workflow command had empty target field"
+                                    .to_string(),
+                            ))
+                        }
+                        Some(sig_we::Target::ChildWorkflowSeq(seq)) => {
+                            let child_wf_machine =
+                                self.get_machine_key(CommandID::ChildWorkflowStart(seq))?;
+                            let mach = self.machine(child_wf_machine);
+                            let as_child: Option<&ChildWorkflowMachine> =
+                                mach.as_any().downcast_ref();
+                            if let Some(child_machine) = as_child {
+                                (
+                                    WorkflowExecution {
+                                        namespace: child_machine.namespace().to_string(),
+                                        workflow_id: child_machine.workflow_id().to_string(),
+                                        run_id: "".to_string(),
+                                    },
+                                    true,
+                                )
+                            } else {
+                                todo!("Return signal failed")
+                            }
```

Comment:
I sorta hate this downcasting. It can be avoided if lang doesn't do the whole sequence number thing for child workflows and instead just passes the namespace/wfid and says "I think it's my child". That would be more like what the other SDKs do.

This is potentially slightly more correct in that we can prevent things we know won't work ahead of time, but it introduces this whole new downcasting path and can fail in it's own interesting ways...

@bergundy WDYT? Should we keep this or just change the command from lang?

Reply:
Yeah but we don't need to downcast in that case because the generic `cancel` method exists

Reply:
I will undo this as we discussed